File: rust/2015/day17b/src/main.rs
use std::env;
use std::fs;
use std::io;
use std::process;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn process_file(filename: &str) -> io::Result<u32> {
    let containers: Vec<u32> = fs::read_to_string(filename)
        .unwrap()
        .lines()
        .map(|line| line.parse::<u32>().unwrap())
        .collect();
    let n = containers.len();
    let mut combo_lens: Vec<u32> = vec![];
    for i in 0..(1 << n) {
        let mut combo_len = 0;
        let mut sum = 0;
        for (j, item) in containers.iter().enumerate().take(n) {
            if (i & (1 << j)) != 0 {
                sum += item;
                combo_len += 1;
            }
        }
        if sum == 150 {
            combo_lens.push(combo_len);
        }
    }
    let min_containers = combo_lens.iter().min().unwrap();
    let min_ways = combo_lens.iter().filter(|x| *x == min_containers).count() as u32;
    Ok(min_ways)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}

File: haskell/2015/day17b/src/Main.hs
module Main (main) where

import Data.Int (Int32)
import Data.List
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Control.DeepSeq
import System.Clock

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

process :: String -> Int32
process content =
    let containers = (map read (lines content)) :: [Int]
        combos = filter (\x -> sum x == 150) (subsequences containers)
        minContainers = minimum $ map length combos
        minWays = filter (\x -> length x == minContainers) combos
    in fromIntegral $ length minWays


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"
main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname
